### PR TITLE
Added namespaces template

### PR DESCRIPTION
**Pull request by alijared**
In order to have subcommands for entities it's necessary to set up namespaces for argh

### CODE REVIEW

File: croud/__main__.py [hunk 1]
```diff
@@ -23,10 +23,13 @@
...
 import argh
 import colorama
-import pkg_resources
 
 from croud import __version__
 from croud.config import Configuration
+from croud.login import login
+from croud.logout import logout
+from croud.me import me
+from croud.projects.list import list
```
Comment (mikethebeer) [inline croud/__main__.py]: would name it to `projects_list` to avoid possible conflict with a keyword

File: croud/me.py [hunk 2]
```diff
@@ -35,7 +35,7 @@
     type=str,
 )
 @argh.arg("-o", "--output-fmt", choices=["json"], default="json", type=str)
-@argh.arg("--env", choices=["prod", "env"], default=None, type=str)
+@argh.arg("--env", choices=["prod", "dev"], default=None, type=str)
```
Comment (mikethebeer) [inline croud/me.py]: this needs to be fixed in a separate hotfix PR

File: croud/config.py [hunk 3]
```diff
@@ -49,7 +49,7 @@ def create() -> None:
 
     @staticmethod
     def get_env() -> str:
-        if not Configuration.current_context:
+        if Configuration.current_context:
```
Comment (mikethebeer) [inline croud/config.py]: hotfix pr

File: croud/projects/list.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    type=str,
+)
+@argh.arg("-o", "--output-fmt", choices=["json"], default="json", type=str)
+@argh.arg("--env", choices=["prod", "dev"], default=None, type=str)
+def list(region=None, output_fmt=None, env=None) -> None:
+    """
+    Lists all projects for the current user in the specified region
+    """
+
+    # Todo: Return list of projects for logged in user in specified region
```
Comment (mikethebeer) [inline croud/projects/list.py]: To be more pythonic 👉 `pass`
```suggestion
    # Todo: Return list of projects for logged in user in specified region
    pass
```

**Comment by mfussenegger** [comment]
Btw. maybe we should consider moving away from `argh`. Based on my experience using it for cr8 I probably wouldn't use it anymore as it's difficult to compose argument definitions.

We're already starting to run into this problem here that we've multiple sub-commands with the same options. For example `@argh.arg("--env", choices=["prod", "dev"], default=None, type=str)`

Going with plain argparse would make it possible to extract the definitions into re-usable functions. E.g. `add_env_option(parser)`. That way we would ensure that the choices, description and defaults are consistent across all commands.

**Comment by mikethebeer** [comment]
@mfussenegger yes, I had already a bad feeling to start with `argh` here. I would say we could already start switching to the builtin `argparse`. @alijared would it be possible that you make the switch to `argparse` here (in this PR) already?

**Comment by alijared** [comment]
@mfussenegger I agree this approach would be better. @mikethebeer I will update this PR

File: croud/__main__.py [hunk 5]
```diff
@@ -21,28 +21,71 @@
...
-    commands = [
-        entry_point.load()
-        for entry_point in pkg_resources.iter_entry_points("croud_commands")
-    ]
-    p.add_commands(commands)
-    p.dispatch()
+    commands: dict = {
+        "me": {
+            "description": "Prints the current logged in user",
+            "usage": "croud me [-h] [--env {prod,dev}]\n\t\t"
```
Comment (mikethebeer) [inline croud/__main__.py]: why `\n\t\t` ?

File: croud/__main__.py [hunk 6]
```diff
@@ -21,28 +21,71 @@
...
+            "[-r {westeurope.azure,eastus.azure,bregenz.a1}] "
+            "[-o {json}]",
+            "extra_args": [region_arg, output_fmt_arg],
+            "calls": me,
+        },
+        "login": {
+            "description": "Performs an OAuth2 Login to CrateDB Cloud",
+            "usage": "croud login [-h] [--env {prod,dev}]",
```
Comment (mikethebeer) [inline croud/__main__.py]: is it possible to extract the usage description of the `default_args` like `--env` to make it reusable?

File: croud/cmd.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+from argparse import ArgumentParser, Namespace
+from typing import Callable
+
+
+class CMD:
+    def __init__(self, commands: dict):
+        parser: ArgumentParser = argparse.ArgumentParser()
+        self.create_parent_cmd(parser, 1, commands)
+
+    def create_parent_cmd(self, parser: ArgumentParser, depth: int, commands: dict):
```
Comment (mikethebeer) [inline croud/cmd.py]: missing tests

File: croud/__main__.py (hunk 5 above)
Comment (alijared) [reply croud/__main__.py]: Formatting. Otherwise the line is very, very long. It's what argparse does for you if you don't specify a custom `usage` field

File: croud/__main__.py (hunk 6 above)
Comment (alijared) [reply croud/__main__.py]: It is, but for some reason argparse doesn't provide the accurate usage, instead it only provides accurate positional and optional arguments. So for login, as an example, the standard usage is `croud [-h] [--env {prod,dev}]`. This doesn't show that the `login` argument is supposed to be specified, which doesn't make sense to me, so I overrode them all

Comment (alijared) [reply croud/__main__.py]: Managed to figure this out. You don't need to specify the usage anymore

File: setup.py [hunk 8]
```diff
@@ -35,15 +35,7 @@
...
-            "env = croud.env:env",
-            "login = croud.login:login",
-            "logout = croud.logout:logout",
-            "me = croud.me:me",
-        ],
-    },
+    entry_points={"console_scripts": ["croud = croud.__main__:main"]},
     packages=["croud"],
     install_requires=[
         "argh",
```
Comment (mikethebeer) [inline setup.py]: can be removed

File: tests/unit_tests/test_commands.py [hunk 9]
```diff
@@ -75,20 +75,7 @@ class TestLogout(unittest.TestCase):
...
         m = mock.mock_open()
         with mock.patch("croud.config.open", m, create=True):
-            logout()
+            logout(Namespace(env="dev"))
 
         mock_set_token.assert_called_once_with("")
         mock_print_info.assert_called_once_with("You have been logged out.")
-
-
-class TestEnv(unittest.TestCase):
```
Comment (mikethebeer) [inline tests/unit_tests/test_commands.py]: why did you remove this test?

Comment (mikethebeer) [reply tests/unit_tests/test_commands.py]: ok understand, thx
